hcl2template/parser.go

```diff
@@ -295,10 +297,224 @@ func filterVarsFromLogs(inputOrLocal Variables) {
 	}
 }
 
+func (cfg *PackerConfig) detectBuildPrereqDependencies() hcl.Diagnostics {
+	var diags hcl.Diagnostics
+
+	for _, ds := range cfg.Datasources {
+		dependencies := GetVarsByType(ds.block, "data")
+		dependencies = append(dependencies, GetVarsByType(ds.block, "local")...)
+
+		for _, dep := range dependencies {
+			// If something is locally aliased as `local` or `data`, we'll falsely
+			// report it as a local variable, which is not necessarily what we
+			// want to process here, so we continue.
+			//
+			// Note: this is kinda brittle, we should understand scopes to accurately
```

Comment:
You mention this is brittle, can you think of a scenario where this would cause valid dependencies to be ignored?

Reply:
So on this, I don't have an example to be fair, I had tried to replicate a concern of mine with local aliases in a `for` expression, but the local aliases weren't returned by the hcl function, so it's mostly a non-concern at this point.
I can remove this comment until we do encounter a real-life example that highlights this problem, but at the present time, it seems somewhat solid on the templates we have in the repo.